examples/cmd/logtransport.go

```diff
@@ -22,24 +23,31 @@ type logTransport struct {
 func (t *logTransport) RoundTrip(r *http.Request) (*http.Response, error) {
 	t0 := time.Now()
 
-	reqBody, _ := ioutil.ReadAll(r.Body)
-	r.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody))
+	backupReqBody, err := httpbody.BackupRequestBody(r)
+	if err != nil {
+		return nil, fmt.Errorf("could not backup the request before sending it: %s", err)
+	}
 
 	resp, err := t.T.RoundTrip(r)
 	if err != nil {
-		return resp, err
+		return nil, err
 	}
 
-	respBody, err := ioutil.ReadAll(resp.Body)
+	reqBody, err := ioutil.ReadAll(backupReqBody)
 	if err != nil {
```

Comment:
why this logging depends on the reading from a buffer that you keep in memory?

Reply:
I don't understand your question.

The `Request` passed by  the argument `r` to this function at [line 22](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R23), is no longer readable after crossing the `RoundTrip(r)` at [line 31](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R31)
So if we want to log the original `Request` at [line 49](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R49), we need a backup from that original `Request` made before `RoundTrip(r)` as already explained by https://github.com/src-d/metadata-retrieval/pull/57#issuecomment-543129407

Reply:
the log line doesn't depend on the content of request or response at all. It reads only `StatusCode` and `URL`.

Reply:
to be more clear `With` should be moved out of the `if`.
And if you return just `[]byte` above you won't need this `if` at all.

Reply:
1. [line 49](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R49) logs the content of the `Request.Body`, so it needes it.
1. since the `Request.Body` is needed at seen per `(1)`, it is read by [line 36](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R36).
1. reading `(2)` the `Request.Body` can fail, so **if** it fails, it logs by [line 40](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R40) `could not log the request: %s`.
1. then, the `if` at [line 37](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R37) is needed to capture the failure when reading `2`.

Reply:
1. [line 49](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R49) logs the content of the `Request.Body`, so it needes it.
1. since the `Request.Body` is needed at seen per `(1)`, it is read by [line 36](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R36).
1. reading `(2)` the `Request.Body` can fail, so **if** it fails, it logs by [line 40](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R40) `could not log the request: %s`.
1. then, the `if` at [line 37](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-b99cf0f6f6715e2e68843770335d7f09R37) is needed to capture the failure when reading `2`.

Reply:
`logtransport` was restored to old version by 254ed27242e8745c89bbd34fd07e4eb09dab7160